toolchain/check/handle_function.cpp

```diff
@@ -297,6 +297,16 @@ static auto BuildFunctionDecl(Context& context,
   // Write the function ID into the FunctionDecl.
   context.ReplaceInstBeforeConstantUse(decl_id, function_decl);
 
+  // Diagnose `abstract` function with definition using the canonical Function's
+  // modifiers.
+  if (is_definition &&
+      context.functions().Get(function_decl.function_id).virtual_modifier ==
+          SemIR::Function::VirtualModifier::Abstract) {
+    CARBON_DIAGNOSTIC(DefinedAbstractFunction, Error,
+                      "`abstract` function with definition");
```

Comment:
I'd find this slightly more readable as "definition of `abstract` function" but no strong preference. (I think the "with" phrasing is closer to what we use elsewhere.)